Panic during ensure dry run

vault-helper panics during an ensure dry run.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x2f817ac]

goroutine 156 [running]:
github.com/jetstack/tarmak/vendor/github.com/jetstack/vault-helper/pkg/kubernetes.(*Kubernetes).ensureDryRunPKIRolesEtcd(0xc420b9b6c0, 0xc4209533b0, 0x41901e0, 0xc420a33630, 0xc4207bdf00)
        /home/christian/.golang/packages/src/github.com/jetstack/tarmak/vendor/github.com/jetstack/vault-helper/pkg/kubernetes/kubernetes_pki_roles.go:189 +0xac
github.com/jetstack/tarmak/vendor/github.com/jetstack/vault-helper/pkg/kubernetes.(*Kubernetes).EnsureDryRun(0xc420b9b6c0, 0x3f93220, 0xc420c74860, 0x0)
        /home/christian/.golang/packages/src/github.com/jetstack/tarmak/vendor/github.com/jetstack/vault-helper/pkg/kubernetes/kubernetes.go:289 +0x199
github.com/jetstack/tarmak/pkg/terraform/providers/tarmak/rpc.(*tarmakRPC).VaultClusterStatus(0xc420c74840, 0xc420188af0, 0xc4201d4970, 0x0, 0x0)
        /home/christian/.golang/packages/src/github.com/jetstack/tarmak/pkg/terraform/providers/tarmak/rpc/vault_cluster_status.go:74 +0x785
reflect.Value.call(0xc420bee0c0, 0xc420e23170, 0x13, 0x3dfe1d7, 0x4, 0xc420092f18, 0x3, 0x3, 0xc420ece240, 0x43d094, ...)
        /home/christian/.golang/go/src/reflect/value.go:447 +0x969
reflect.Value.Call(0xc420bee0c0, 0xc420e23170, 0x13, 0xc42018ff18, 0x3, 0x3, 0xc420a0d810, 0xffffffffffffffff, 0xc42018ffb0)
        /home/christian/.golang/go/src/reflect/value.go:308 +0xa4
net/rpc.(*service).call(0xc4207e2f80, 0xc420937310, 0xc420a0c010, 0xc420a0c030, 0xc420224300, 0xc420552000, 0x3221820, 0xc420188af0, 0x16, 0x3221860, ...)
        /home/christian/.golang/go/src/net/rpc/server.go:384 +0x14e
created by net/rpc.(*Server).ServeCodec
        /home/christian/.golang/go/src/net/rpc/server.go:480 +0x43a
```


/assign

--- a/pkg/kubernetes/kubernetes_fake_test.go
+++ b/pkg/kubernetes/kubernetes_fake_test.go
@@ -2,6 +2,7 @@
 package kubernetes
 
 import (
+	"errors"
 	"fmt"
 	"testing"
 
@@ -180,4 +181,9 @@ func (v *fakeVault) PKIEnsure() {
 
 	first := v.fakeSys.EXPECT().Mount("test-cluster-inside/pki/wrong-type-pki", gomock.Any()).Times(1).Return(nil)
 	v.fakeSys.EXPECT().Mount("test-cluster-inside/pki/wrong-type-pki", gomock.Any()).Times(1).Return(fmt.Errorf("wrong type")).After(first)
+}
+
+func (v *fakeVault) ReadPKIRoleErr() {
+	v.fakeVault.EXPECT().Logical().AnyTimes().Return(v.fakeLogical)
+	v.fakeLogical.EXPECT().Read(gomock.Any()).Return(nil, errors.New("an error")).AnyTimes()
 }

--- a/pkg/kubernetes/kubernetes_pki_roles.go
+++ b/pkg/kubernetes/kubernetes_pki_roles.go
@@ -148,22 +148,21 @@ func (k *Kubernetes) k8sComponentRole(roleName string) *pkiRole {
 
 // this makes sure all etcd PKI roles are setup correctly
 func (k *Kubernetes) ensurePKIRolesEtcd(p *PKIVaultBackend) error {
-	var err error
-	var result error
+	var result *multierror.Error
 
-	if err = p.WriteRole(k.etcdClientRole()); err != nil {
+	if err := p.WriteRole(k.etcdClientRole()); err != nil {
 		result = multierror.Append(result, err)
 	}
 
-	if err = p.WriteRole(k.etcdServerRole()); err != nil {
+	if err := p.WriteRole(k.etcdServerRole()); err != nil {
 		result = multierror.Append(result, err)
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) deletePKIRolesEtcd(p *PKIVaultBackend) error {
-	var result error
+	var result *multierror.Error
 
 	if err := p.DeleteRole(k.etcdClientRole()); err != nil {
 		result = multierror.Append(result, err)
@@ -173,7 +172,7 @@ func (k *Kubernetes) deletePKIRolesEtcd(p *PKIVaultBackend) error {
 		result = multierror.Append(result, err)
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) ensureDryRunPKIRolesEtcd(p *PKIVaultBackend) (bool, error) {
@@ -182,8 +181,10 @@ func (k *Kubernetes) ensureDryRunPKIRolesEtcd(p *PKIVaultBackend) (bool, error) 
 	secret, err := p.ReadRole(k.etcdClientRole())
 	if err != nil {
 		result = multierror.Append(result, err)
-	} else if secret == nil || secret.Data == nil || len(secret.Data) == 0 {
-		return true, nil
+	}
+
+	if secret == nil || secret.Data == nil || len(secret.Data) == 0 {
+		return true, result.ErrorOrNil()
 	}
 
 	if !secretDataMatch(secret.Data, k.etcdClientRole().Data) {
@@ -193,8 +194,10 @@ func (k *Kubernetes) ensureDryRunPKIRolesEtcd(p *PKIVaultBackend) (bool, error) 
 	secret, err = p.ReadRole(k.etcdServerRole())
 	if err != nil {
 		result = multierror.Append(result, err)
-	} else if len(secret.Data) == 0 {
-		return true, nil
+	}
+
+	if secret == nil || len(secret.Data) == 0 {
+		return true, result.ErrorOrNil()
 	}
 
 	if !secretDataMatch(secret.Data, k.etcdServerRole().Data) {
@@ -217,7 +220,7 @@ func (k *Kubernetes) pkiRoleK8s() []*pkiRole {
 
 // this makes sure all kubernetes PKI roles are setup correctly
 func (k *Kubernetes) ensurePKIRolesK8S(p *PKIVaultBackend) error {
-	var result error
+	var result *multierror.Error
 
 	for _, role := range k.pkiRoleK8s() {
 		if err := p.WriteRole(role); err != nil {
@@ -225,11 +228,11 @@ func (k *Kubernetes) ensurePKIRolesK8S(p *PKIVaultBackend) error {
 		}
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) deletePKIRolesK8S(p *PKIVaultBackend) error {
-	var result error
+	var result *multierror.Error
 
 	for _, role := range k.pkiRoleK8s() {
 		if err := p.DeleteRole(role); err != nil {
@@ -237,7 +240,7 @@ func (k *Kubernetes) deletePKIRolesK8S(p *PKIVaultBackend) error {
 		}
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) ensureDryRunPKIRolesK8S(p *PKIVaultBackend) (bool, error) {
@@ -247,8 +250,10 @@ func (k *Kubernetes) ensureDryRunPKIRolesK8S(p *PKIVaultBackend) (bool, error) {
 		secret, err := p.ReadRole(role)
 		if err != nil {
 			result = multierror.Append(result, err)
-		} else if len(secret.Data) == 0 {
-			return true, nil
+		}
+
+		if secret == nil || len(secret.Data) == 0 {
+			return true, result.ErrorOrNil()
 		}
 
 		if !secretDataMatch(secret.Data, role.Data) {
@@ -261,7 +266,7 @@ func (k *Kubernetes) ensureDryRunPKIRolesK8S(p *PKIVaultBackend) (bool, error) {
 
 // this makes sure all kubernetes API Proxy PKI roles are setup correctly
 func (k *Kubernetes) ensurePKIRolesK8SAPIProxy(p *PKIVaultBackend) error {
-	var result error
+	var result *multierror.Error
 
 	roles := []*pkiRole{
 		k.k8sAPIServerProxyRole(),
@@ -273,11 +278,11 @@ func (k *Kubernetes) ensurePKIRolesK8SAPIProxy(p *PKIVaultBackend) error {
 		}
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) deletePKIRolesK8SAPIProxy(p *PKIVaultBackend) error {
-	var result error
+	var result *multierror.Error
 
 	roles := []*pkiRole{
 		k.k8sAPIServerProxyRole(),
@@ -289,24 +294,20 @@ func (k *Kubernetes) deletePKIRolesK8SAPIProxy(p *PKIVaultBackend) error {
 		}
 	}
 
-	return result
+	return result.ErrorOrNil()
 }
 
 func (k *Kubernetes) ensureDryRunPKIRolesK8SAPIProxy(p *PKIVaultBackend) (bool, error) {
 	secret, err := p.ReadRole(k.k8sAPIServerProxyRole())
-	if err != nil {
-		return false, err
-	}
-
-	if len(secret.Data) == 0 {
-		return true, nil
+	if secret == nil || len(secret.Data) == 0 {
+		return true, err
 	}
 
 	if !secretDataMatch(secret.Data, k.k8sAPIServerProxyRole().Data) {
-		return true, nil
+		return true, err
 	}
 
-	return false, nil
+	return false, err
 }
 
 func constructTimeString(t time.Duration) string {

--- a/pkg/kubernetes/pki_vault_backend_test.go
+++ b/pkg/kubernetes/pki_vault_backend_test.go
@@ -77,3 +77,34 @@ func TestPKIVaultBackend_Ensure(t *testing.T) {
 		t.Errorf("expected an error from wrong pki type")
 	}
 }
+
+func TestPKIVaultBackend_DryRunErr(t *testing.T) {
+	fv := NewFakeVault(t)
+	defer fv.Finish()
+
+	fv.ExpectWrite()
+	fv.ReadPKIRoleErr()
+
+	fk := fv.Kubernetes()
+	pki := NewPKIVaultBackend(fk, "wrong-type-pki", logrus.NewEntry(logrus.New()))
+
+	type pkiFuncs struct {
+		f func(*PKIVaultBackend) (bool, error)
+		s string
+	}
+
+	for _, p := range []pkiFuncs{
+		{fk.ensureDryRunPKIRolesEtcd, "Etcd"},
+		{fk.ensureDryRunPKIRolesK8S, "K8s"},
+		{fk.ensureDryRunPKIRolesK8SAPIProxy, "K8sApiProxy"},
+	} {
+		changeNeeded, err := p.f(pki)
+		if err == nil {
+			t.Errorf("expected error, got none (%s)", p.s)
+		}
+
+		if !changeNeeded {
+			t.Errorf("expected change needed, got none (%s)", p.s)
+		}
+	}
+}

